src/librustc_lint/types.rs

```diff
@@ -428,41 +428,47 @@ enum FfiResult<'tcx> {
 }
 
 /// Check if this enum can be safely exported based on the
-/// "nullable pointer optimization". Currently restricted
-/// to function pointers and references, but could be
-/// expanded to cover NonZero raw pointers and newtypes.
+/// "nullable pointer optimization".
 /// FIXME: This duplicates code in codegen.
-fn is_repr_nullable_ptr<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
-                                  def: &'tcx ty::AdtDef,
+fn is_repr_nullable_ptr<'a, 'tcx>(visitor: &ImproperCTypesVisitor<'a, 'tcx>,
+                                  cache: &mut FxHashSet<Ty<'tcx>>,
+                                  tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                  ty: Ty<'tcx>,
+                                  ty_def: &'tcx ty::AdtDef,
                                   substs: SubstsRef<'tcx>)
                                   -> bool {
-    if def.variants.len() == 2 {
-        let data_idx;
+    if ty_def.variants.len() != 2 {
+        return false;
+    }
 
-        let zero = VariantIdx::new(0);
-        let one = VariantIdx::new(1);
+    let data_idx;
 
-        if def.variants[zero].fields.is_empty() {
-            data_idx = one;
-        } else if def.variants[one].fields.is_empty() {
-            data_idx = zero;
-        } else {
-            return false;
-        }
+    let zero = VariantIdx::new(0);
+    let one = VariantIdx::new(1);
 
-        if def.variants[data_idx].fields.len() == 1 {
-            match def.variants[data_idx].fields[0].ty(tcx, substs).sty {
-                ty::FnPtr(_) => {
-                    return true;
-                }
-                ty::Ref(..) => {
-                    return true;
-                }
-                _ => {}
-            }
-        }
+    if ty_def.variants[zero].fields.is_empty() {
```

Comment:
Rewriting as `let data_idx = if ... { one } else if ... { zero } else { return false; }` would be more readable here.